src/nxt_otel.c

```diff
@@ -0,0 +1,356 @@
+
+/*
+ * Copyright (C) F5, Inc.
+ */
+
+#include <math.h>
+
+#include <nxt_router.h>
+#include <nxt_http.h>
+#include <nxt_otel.h>
+#include <nxt_mp.h>
+#include <nxt_work_queue.h>
+#include <nxt_main.h>
+#include <nxt_conf.h>
+#include <nxt_types.h>
+#include <nxt_string.h>
+
+
+#define NXT_OTEL_TRACEPARENT_LEN 55
+#define NXT_OTEL_BODY_SIZE_TAG "body size"
+#define NXT_OTEL_METHOD_TAG "method"
+#define NXT_OTEL_PATH_TAG "path"
+
+
+static inline void
+nxt_otel_state_transition(nxt_otel_state_t *state, nxt_otel_status_t status)
+{
+    if (status == NXT_OTEL_ERROR_STATE || state->status != NXT_OTEL_ERROR_STATE) {
+        state->status = status;
+    }
+}
+
+
+static inline void
+nxt_otel_trace_and_span_init(nxt_task_t *t, nxt_http_request_t *r)
+{
+    r->otel->trace =
+        nxt_otel_get_or_create_trace(r->otel->trace_id);
+    if (r->otel->trace == NULL) {
+        nxt_log(t, NXT_LOG_ERR, "error generating otel span");
+        nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+        return;
+    }
+
+    nxt_otel_state_transition(r->otel, NXT_OTEL_HEADER_STATE);
+}
+
+
+static void
+nxt_otel_span_add_headers(nxt_task_t *t, nxt_http_request_t *r)
+{
+    nxt_http_field_t *f, *cur;
+    u_char *traceval, *name_cur, *val_cur;
+
+    nxt_log(t, NXT_LOG_DEBUG, "adding headers to trace");
+
+    if (r->otel == NULL || r->otel->trace == NULL)
+    {
+        nxt_log(t, NXT_LOG_ERR, "no trace to add events to!");
+        nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+        return;
+    }
+
+    nxt_list_each(cur, r->fields) {
+        // we need this in a continguous and null terminated segment of memory for Rust FFI
+        name_cur = nxt_mp_zalloc(r->mem_pool, cur->name_length + 1);
+        val_cur = nxt_mp_zalloc(r->mem_pool, cur->value_length + 1);
+        if (name_cur != NULL && val_cur != NULL) {
+            nxt_cpystrn(name_cur, cur->name, cur->name_length);
+            nxt_cpystrn(val_cur, cur->value, cur->value_length);
+            nxt_otel_add_event_to_trace(r->otel->trace, name_cur, val_cur);
+        }
```

Comment:
This is still bugging me...

At the very least we should use @hongzhidao 's suggestion.

However this does beg the question of memory allocation lifetimes.

What will stop this memory being free'd before the rust side of things has finished with it?

Reply:
> What will stop this memory being free'd before the rust side of things has finished with it?

We ran into that, so we copy the value out of the pointer rather than pass the pointer to the batch exporter.

If the memory is freed, the value behind the pointer is `UUUUUUUUUUU...`, which is binary `01010101 01010101 01010101 ...`, which isn't what we want.

Reply:
I don't know whether we can do that, because the Rust side doesn't know about the type declarations in the C side. 🤔

Reply:
That shouldn't be an issue, for example `wasm-wasi-component` has [this](https://github.com/nginx/unit/blob/24ed91f40634372d99f67f0e4e3c2ac0abde81bd/src/wasm-wasi-component/src/lib.rs#L28)

```rust
    pub const fn nxt_string(s: &'static str) -> nxt_str_t {                     
        nxt_str_t {                                                             
            start: s.as_ptr().cast_mut(),                                       
            length: s.len(),                                                    
        }                                                                       
    } 
```

Or in libunit-wasm I [have](https://github.com/nginx/unit-wasm/blob/4d4f1f2284d54dfabf58246034b8fc7fd0282f83/src/rust/unit-wasm-sys/rusty.rs#L39) which is a rust version of [this](https://github.com/nginx/unit-wasm/blob/4d4f1f2284d54dfabf58246034b8fc7fd0282f83/src/c/include/unit/unit-wasm.h#L149).

So definitely doable...

Reply:
Totally butchered example, but you get the idea...

```rust
// lib.rs

use std::slice;
use std::str;

#[repr(C)]
pub struct str_t {
    pub start: *const u8,
    pub len: usize,
}

#[no_mangle]
pub unsafe extern "C" fn hello_from_rust(c_struct: *const str_t) {
    let str_struct = &*c_struct;
    let slice = slice::from_raw_parts(str_struct.start, str_struct.len);
    let s = str::from_utf8(slice).unwrap();

    println!("{}", s);
}
```

```c
/* rusty.c */

#include <stdio.h>
#include <stdint.h>
#include <string.h>

typedef struct {
	uint8_t *start;
	size_t len;
} str_t;

extern void hello_from_rust(const str_t *s);

int main(void)
{
	str_t s = { .start = (uint8_t *)"Hello from C via Rust!",
		    .len = strlen("Hello from C via Rust!") };

	hello_from_rust(&s);

	return 0;
}
```

```
$ LD_LIBRARY_PATH=rusty-lib/target/debug ./rusty
Hello from C via Rust!
```

Reply:
So to take it full circle, this code

```c
    nxt_list_each(cur, r->fields) {
        // we need this in a continguous and null terminated segment of memory for Rust FFI
        name_cur = nxt_mp_zalloc(r->mem_pool, cur->name_length + 1);
        val_cur = nxt_mp_zalloc(r->mem_pool, cur->value_length + 1);
        if (name_cur && val_cur) {
            strncpy((char *) name_cur, (char *) cur->name, cur->name_length);
            strncpy((char *) val_cur, (char *) cur->value, cur->value_length);
            nxt_otel_add_event_to_trace(r->otel->trace, name_cur, val_cur);
        }
    } nxt_list_loop;
```

Would become something like

```c
    nxt_list_each(cur, r->fields) {
        nxt_str_t  name, val;
        
        name = (nxt_str_t){ .start = cur->name, .len = cur->name_length };
        val = (nxt_str_t){ .start = cur->value, .len = cur->value_length };
        
        nxt_otel_add_event_to_trace(r->otel->trace, &name, &val);
    } nxt_list_loop
```

Wowzers!